library/core/src/slice/iter.rs

```diff
@@ -2222,82 +2219,70 @@ impl<'a, T, const N: usize> Iterator for ArrayWindows<'a, T, N> {
 
     #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        if self.num == 0 {
-            return None;
+        if let Some(ret) = self.v.first_chunk() {
+            self.v = &self.v[1..];
+            Some(ret)
+        } else {
+            None
         }
-        // SAFETY:
-        // This is safe because it's indexing into a slice guaranteed to be length > N.
-        let ret = unsafe { &*self.slice_head.cast::<[T; N]>() };
-        // SAFETY: Guaranteed that there are at least 1 item remaining otherwise
-        // earlier branch would've been hit
-        self.slice_head = unsafe { self.slice_head.add(1) };
-
-        self.num -= 1;
-        Some(ret)
     }
 
     #[inline]
     fn size_hint(&self) -> (usize, Option<usize>) {
-        (self.num, Some(self.num))
+        if self.v.len() < N {
+            (0, Some(0))
+        } else {
+            let size = self.v.len() - N + 1;
+            (size, Some(size))
+        }
     }
 
     #[inline]
     fn count(self) -> usize {
-        self.num
+        self.size_hint().0
```

Comment:
I'd probably reverse the dependence here. With count being `self.v.len().saturating_sub(N - 1)` and size_hint giving `(count(), Some(count()))`.

Reply:
`count` takes `self`, but `size_hint` only `&self`.

Reply:
s/self.len()/self.size_hint/ ?